src/components/workspace/editorMode/source.js

```diff
@@ -0,0 +1,831 @@
+ }
+                }
+                
+                return range;
+            }
+    
+            if (foldStyle === "markbegin") {
+                return;
+            }
+    
+            match = line.match(this.foldingStopMarker);
+            if (match) {
+                i = match.index + match[0].length;
+    
+                if (match[1]) {
+                    return this.closingBracketBlock(session, match[1], row, i);
+                }
+    
+                return session.getCommentFoldRange(row, i, -1);
+            }
+        };
+        
+        this.getSectionRange = function(session, row) {
+            var line = session.getLine(row);
+            var startIndent = line.search(/\S/);
+            var startRow = row;
+            var startColumn = line.length;
+            row = row + 1;
+            var endRow = row;
+            var maxRow = session.getLength();
+            while (++row < maxRow) {
+                line = session.getLine(row);
+                var indent = line.search(/\S/);
+                if (indent === -1) {
+                    continue;
+                }
+                if  (startIndent > indent) {
+                    break;
+                }
+                var subRange = this.getFoldWidgetRange(session, "all", row);
+                
+                if (subRange) {
+                    if (subRange.start.row <= startRow) {
+                        break;
+                    } else if (subRange.isMultiLine()) {
+                        row = subRange.end.row;
+                    } else if (startIndent === indent) {
+                        break;
+                    }
+                }
+                endRow = row;
+            }
+            
+            return new Range(startRow, startColumn, endRow, session.getLine(endRow).length);
+        };
+        this.getCommentRegionBlock = function(session, line, row) {
+            var startColumn = line.search(/\s*$/);
+            var maxRow = session.getLength();
+            var startRow = row;
+            
+            var re = /^\s*(?:\/\*|\/\/|--)#?(end)?region\b/;
+            var depth = 1;
+            while (++row < maxRow) {
+                line = session.getLine(row);
+                var m = re.exec(line);
+                if (!m) { continue; }
+                if (m[1]) { depth--; }
+                else { depth++; }
+    
+                if (!depth) { break; }
+            }
+    
+            var endRow = row;
+            if (endRow > startRow) {
+                return new Range(startRow, startColumn, endRow, line.length);
+            }
+        };
+    
+    }).call(FoldMode.prototype);
+    
+    });
+    
+    ace.define("ace/mode/javascript",["require","exports","module","ace/lib/oop","ace/mode/text","ace/mode/javascript_highlight_rules","ace/mode/matching_brace_outdent","ace/worker/worker_client","ace/mode/behaviour/cstyle","ace/mode/folding/cstyle"], function(acequire, exports, module) {
```

Comment:
This looks like another long line with many String literals. Could we shorten this line, and put the Strings somewhere else as constants? Ideally, the function would exist outside as well (to me, this kind of "one-line" programming is really difficult to read)

Reply:
Yep changed it.